### ISSUE TITLE
[css-text] 'Punctuation space' handling for non-CJK languages

### ISSUE THREAD

**Issue by r12a** [w3c/csswg-drafts#8661]
I've been trying to get my head around the ramifications of spaces before punctuation for a while now, because i think the implications go well beyond the case of French, but i still need to marshal more evidence. However, here are some thoughts.

To be honest, I'm inclined, at this point, to wonder whether this should be disunified from the autospace property. Ways in which this punctuation spacing differ conceptually from CJK autospace include the following:

1. punctuation spacing does not add balanced spaces around embedded items (which are from a different script)
2. punctuation spacing usually involves a requirement not to wrap the punctuation mark separately onto the following line during text wrap
3. the size of the gap is not the same as for autospacing, and may vary according to the script (or possibly the punctuation mark) in question

Here are some examples where punctuation marks may need to be separated by a gap from the preceding text. The examples are taken from a cursory browse through my orthography notes.

- N'Ko text may contain commas that are separated by a gap from the end of the previous word: see https://r12a.github.io/scripts/nkoo/nqo.html?showIndex#phrase
- Newar, ditto for danda and double danda: see https://r12a.github.io/scripts/newa/new.html?showIndex#phrase
- Odia, ditto for danda and double danda (to avoid serious ambiguity): see https://r12a.github.io/scripts/orya/or.html?showIndex#phrase
- Adlam text appears to separate most punctuation marks from preceding text: see https://r12a.github.io/scripts/adlm/fuf.html?showIndex#phrase
- Punjabi danda: see https://r12a.github.io/scripts/guru/pa.html?showIndex#phrase
- Santali danda & double danda: see https://r12a.github.io/scripts/olck/sat.html?showIndex#phrase
- Thai repetition mark is optionally preceded by a fixed width, non breaking gap: see https://r12a.github.io/scripts/thai/th.html?showIndex#phrase
- Tibetan often has a gap before or between punctuation marks: see https://r12a.github.io/scripts/tibt/bo.html?showIndex#phrase
- I have also come across Hindi and Mongolian texts where ordinary spaces have been used to move sentence-final punctuation away from the preceding text. 

Then there are characters with a PR line-break property, which signals that they should not be wrapped when alongside a sequence of digits, even if separated by a space. I suspect that for several of these, the space may depend on the preference of the content author, or may benefit from a CSS directive that can standardised or change the approach taken throughout a document.

The size of the gap will vary from language to language, and maybe punctuation mark to punctuation mark within a given language.  But browsers may also apply different gap sizes, if they feel that their size better meets the needs of their users.

The replace function will still be valuable. In some cases a gap may be built into the font glyph for the punctuation, in other cases it is created by the author inserting a regular space. In other use cases, no gap may be built in to the font but application of an appropriate gap varies from content author to content author, or may even vary within the same document. A CSS property that produces standardised gaps across a document, where needed depending on the font or the author preferences, would seem like a useful thing to have.  If that property could also prompt a user agent to apply intelligence to establish the correct sized gap and wrapping behaviour for a given punctuation mark or symbol for a given language (there probably aren't that many to consider) that would be an additional benefit.

In summary:

[1] should this behaviour be bundled with the somewhat different CJK autospace behaviour?

[2] perhaps we should edit the current spec text at https://www.w3.org/TR/css-text-4/#valdef-text-autospace-punctuation to make it clear that French is just one possible application among many, and make the case that browsers can begin addressing the needs of other languages right now without waiting for 'future specifications'. I'm not sure whether there needs to be a registry for such information.

**Comment by fantasai** [comment]
> I'm not sure that there needs to be a registry for such information.

We are going to want interoperability here, so I think we do need a registry. I don't mind if we want to maintain one here at W3C (we've got a Registry track now!), or if we want to ask Unicode to spin something up, but if an implementation wants to add a rule they should have to register it so we can all add the same rules.

**Comment by r12a** [comment]
I'm not convinced that we do need interoperability in terms of what the specific width of the gap will be, nor that it will be easy to get information and arrive at a consensus on what it should be for each and every language.  That kind of thing can vary by font already, and browsers should probably be allowed to experiment a little, especially for scripts that have little available documentation about this.

Bear in mind, also that the gap width preceding a danda may not be the same gap width needed before another character that is separated from the foregoing text by a gap. And that sometimes different users may currently use different widths in the same context.

There are already plenty of things that we leave to the discretion of the implementers, and to me this seems like another that would be appropriate for that.

**Comment by xfq** [comment]
We also [discussed](https://github.com/w3c/csswg-drafts/issues/8661) this issue at the Chinese Layout Task Force Teleconference this week, and there was also comments about we should not this bundle this behaviour with the CJK autospace behaviour, because the French case is an issue of grammar or writing style, while ideograph and non-ideograph case is a design issue.

We also agree that we should use language tags to achieve better results.

re interoperability, if we don't require interoperability, I'm a bit worried there will be a problem of one line becoming two lines in different browsers for shorter texts like a label or some button text. But maybe there won't be much difference between browsers and I'm worrying too much.

**Comment by r12a** [comment]
> re interoperability, if we don't require interoperability, I'm a bit worried there will be a problem of one line becoming two lines in different browsers for shorter texts like a label or some button text. But maybe there won't be much difference among browsers and I'm worrying too much.

First, this kind of punctuation space is not very likely to appear in buttons. Second, this kind of resizing happens all the time unless the font in the button is strapped down tight, because of fallback substitutions. Third, we're talking about very small spaces, and (a) if normal spaces have been used the line width should shrink, and (b) if such a small difference causes problems, then the page is not designed well enough to support translation.  So i'm not particularly worried about this.

**Comment by Crissov** [comment]
Is the spacing and breaking before and after En Dash `–` and Em Dash `—` in scope of this issue?
